### ISSUE TITLE
Overriding configuration using an environment doesn't trigger type checks when merging into Structured config

### ISSUE THREAD

**Issue by SimonStJG** [omry/omegaconf#488]
**Describe the bug**

If I create a structured configuration with an integer field and then merge in a configuration where that fields value is taken from an environment variable, the type checks usually associated with the structured field are not triggered.

**To Reproduce**

```
def test_strange_overriding_behaviour():
    @dataclass
    class Base:
        foo: int = omegaconf.MISSING

    override = textwrap.dedent("""
        foo: ${env:FOO_OVERRIDE}
    """)

    with MonkeyPatch.context() as mp:
        mp.setenv("FOO_OVERRIDE", "This is not an integer")
        base_conf = omegaconf.OmegaConf.structured(Base)
        override_conf = omegaconf.OmegaConf.create(override)

        # This does not raise ValidationError but merges successfully, and merged_conf.foo is set to the string
        # "This is not an integer"
        with pytest.raises(omegaconf.ValidationError):
            merged_conf = omegaconf.OmegaConf.merge(base_conf, override_conf)

```

**Expected behavior**

I expected a validation error to be raised on merge, as it would have been had I set:
`override = "foo: "This is not an integer"`

**Additional context**
- [x] OmegaConf version: 2.0.5
- [x] Python version: 3.8.5
- [x] Operating system : Ubuntu 16.04.7 LTS
- [x] Please provide a minimal repro

**Comment by SimonStJG** [comment]
Also tested on 2.0.6

**Comment by omry** [comment]
This is intentional.
Custom resolvers return value is not merged into the config. They can actually return anything (including incompatible objects).

**Comment by omry** [comment]
```python
from dataclasses import dataclass
from omegaconf import SI, OmegaConf

OmegaConf.register_resolver("foo", lambda: "not an int")


@dataclass
class Foo:
    bar: int = SI("${foo:}")


foo = OmegaConf.structured(Foo)
print(foo.bar)
```

**Comment by SimonStJG** [comment]
Hi @omry, thanks for the explanation, that makes sense. Would you be interested in a little docs update pr, perhaps in the Interpolation subsection of the structured config section, to make this behaviour clear?

Also thanks for maintaining this, its been a lovely library to use.

**Comment by odelalleau** [comment]
Reopening as this will be fixed by #445 in OmegaConf 2.1
